### DISCUSSION TITLE
Strange behaviour on generics

### DISCUSSION THREAD

**Discussion by KeyserDSoze** [dotnet/csharplang#5245] [Q&A]
I have this situation
a class
```
 public class MyDocument : IConfigurableDocument
  {
      public string PrimaryKey { get; set; }
      public string SecondaryKey { get; set; }
      public MiniSample Mini { get; set; }
      public DateTime Timestamp { get; set; }

      public RystemDocumentServiceProvider Configure(string callerName)
      {
          return this.StartConfiguration()
              .WithAzure()
              .WithTableStorage(new Azure.Integration.Storage.TableStorageConfiguration(callerName))
              .WithPrimaryKey(x => x.PrimaryKey)
              .WithSecondaryKey(x => x.SecondaryKey)
              .WithTimestamp(x => x.Timestamp)
              .ConfigureAfterBuild();
      }
  }
```
a class that extends this class
```
public class MyRealDocument : MyDocument
{
}
```
In the method Configure() there is an extension method StartConfiguration()
```
public static class ConfigurableDocumentExtensions
{
    public static RystemDocumentServiceProvider<T> StartConfiguration<T>(this T configurableDocument)
        where T : IConfigurableDocument
        => RystemDocumentServiceProvider.Configure(configurableDocument);
}
```
And in the extension method I'm using a static method that creates an instance of RystemDocumentServiceProvider<T>
```
public sealed class RystemDocumentServiceProvider
  {
      private RystemDocumentServiceProvider() { }
      internal static readonly RystemDocumentServiceProvider Instance = new();
      public static RystemDocumentServiceProvider<T> Configure<T>(T entity) 
          => new(ServiceLocatorAtRuntime.PrepareToAddNewService());
  }
```
Ok, now, when I call the Configure() method in this way.
```
new MyRealDocument().Configure()
```
I supposed to have a RystemDocumentServiceProvider<T> as RystemDocumentServiceProvider<MyRealDocument>, when code was running I found instead that that object was a RystemDocumentServiceProvider<MyDocument> even though if I put a breakpoint in 
```
public static RystemDocumentServiceProvider<T> Configure<T>(T entity) 
            => new(ServiceLocatorAtRuntime.PrepareToAddNewService());
```
and I saw that my T entity was a MyRealDocument and not a MyDocument.

![image](https://user-images.githubusercontent.com/40713438/136051520-42d7460d-c355-4b7b-9556-f22b58b2d159.png)

Why do I receive that object as MyDocument and not as MyRealDocument?
Is it a normal behaviour or is it strange for something?

**Comment by svick** [comment]
I'd say this is the expected behavior. You call `this.StartConfiguration()` inside `MyDocument`, so that call is resolved as `ConfigurableDocumentExtensions.StartConfiguration<MyDocument>(this)`. It doesn't matter here that the runtime type of `this` is actually `MyRealDocument`, because its compile-time type, and thus also the `T` in `StartConfiguration`, is still going to be `MyDocument`.

**Reply by KeyserDSoze** [reply]
ok I thought the generic was generated at runtime. Then, it's the correct behaviour. And so, C# creates during compiling all the possible classes to fit. Good news for me. How can I solve my problem, exists a  best practice or a work-around?

I tried to use reflection (I already knew that the returned type would have been the same T) but I tried.
```
public static RystemDocumentServiceProvider<T> Configure<T>(T entity)
{
    var type = typeof(RystemDocumentServiceProvider<>);
    var genericType = type.MakeGenericType(entity.GetType());
    var constructor = genericType.GetConstructor(new Type[1] { typeof(IServiceCollection) });
    var instance = constructor.Invoke(new object[1] { ServiceLocatorAtRuntime.PrepareToAddNewService() });
    return (RystemDocumentServiceProvider<T>)instance;
}
```

**Reply by svick** [reply]
@KeyserDSoze You don't say what is the problem with the reflection code, but I think it won't work, since you're casting an instance of `RystemDocumentServiceProvider<MyRealDocument>` to `RystemDocumentServiceProvider<MyDocument>`, which are two incompatible types. One solution to that would be to create and use a covariant `interface IRystemDocumentServiceProvider<out T>`, if that makes sense for your code.

**Reply by KeyserDSoze** [reply]
no no it works fine, but it returns obviously `RystemDocumentServiceProvider<MyDocument>` but it works without any exception. I don't need to create a covariant interface. In a few words, I need to return `RystemDocumentServiceProvider<MyRealDocument>` in any way passing for a method in MyDocument. Is there any chance to do it somehow?

**Reply by Joe4evr** [reply]
No, only interfaces and delegates can have co-/contravariant type parameters.

**Reply by KeyserDSoze** [reply]
Sorry @svick @Joe4evr I realized what you said only now. Yes you are right, to work with it I used an interface, it's the only way to run that code, but the point is not that, the point is: "Is there any chance to have  `IRystemDocumentServiceProvider<MyRealDocument>` and not  `IRystemDocumentServiceProvider<MyDocument>`?"
